src/events/upgrade.py

```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Upgrades implementation."""
+import logging
+from functools import cached_property
+from typing import TYPE_CHECKING
+
+from charms.data_platform_libs.v0.upgrade import (
+    ClusterNotReadyError,
+    DataUpgrade,
+    DependencyModel,
+    KubernetesClientError,
+)
+from charms.zookeeper.v0.client import QuorumLeaderNotFoundError, ZooKeeperManager
+from kazoo.client import ConnectionClosedError
+from lightkube.core.client import Client
+from lightkube.core.exceptions import ApiError
+from lightkube.resources.apps_v1 import StatefulSet
+from ops.framework import EventBase
+from ops.model import ModelError
+from pydantic import BaseModel
+from tenacity import retry, stop_after_attempt, wait_random
+from typing_extensions import override
+
+from literals import CLIENT_PORT, CONTAINER
+
+if TYPE_CHECKING:
+    from charm import ZooKeeperCharm
+
+logger = logging.getLogger(__name__)
+
+
+class ZooKeeperDependencyModel(BaseModel):
+    """Model for ZooKeeper Operator dependencies."""
+
+    service: DependencyModel
+
+
+class ZKUpgradeEvents(DataUpgrade):
+    """Implementation of :class:`DataUpgrade` overrides for in-place upgrades."""
+
+    def __init__(self, charm: "ZooKeeperCharm", **kwargs):
+        super().__init__(charm, **kwargs)
+        self.charm = charm
+        self.substrate = self.charm.state.substrate
+
+        self.framework.observe(
+            getattr(self.charm.on, "upgrade_charm"), self._on_zookeeper_pebble_ready_upgrade
+        )
+
+    def _on_zookeeper_pebble_ready_upgrade(self, _: EventBase) -> None:
+        """Handler for the `upgrade-charm` events handled during in-place upgrades."""
+        # ensure pebble-ready only fires after normal peer-relation-driven server init
+        if not self.charm.workload.alive or not self.charm.state.unit_server.started or self.idle:
+            return
+
+        try:
+            if self.charm.workload.healthy:
+                self.set_unit_completed()
```

Comment:
Why is the `unit_completed` being called before `init_server`?

Reply:
I think it's because it **might** have self-healed. I don't think it actually will though, but just in case it comes up later on during our dev, that's why it's there.

Reply:
Uhm, I also find it a bit weird and confusing. Can we just not do this? Would it be fair - even if the unit comes back online and self-heal, just do a init_server? I mean, even if it came back online, maybe the upgrade would trigger some updates of the server configuration. It seems a bit dangerous to just skip the init_server JUST because it is online

Also, What is the order of the events here? upgrade first and then pebble ready, or the other way around? Because if the

BTW, this seems a very general use-case, which is not Kafka specific. What people on PostgreSQL and MySQL do in these cases?

Reply:
Maybe it's a good point. I'll remove it.

On the event order, for K8s it goes:
1. upgrade-charm
2. pebble-ready
3. start